collector/container.go

```diff
@@ -0,0 +1,258 @@
+			"Dropped Incoming Packets on Interface",
+			[]string{"container_id", "interface"},
+			nil,
+		),
+		DroppedPacketsOutgoing: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "network_transmit_packets_dropped_total"),
+			"Dropped Outgoing Packets on Interface",
+			[]string{"container_id", "interface"},
+			nil,
+		),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *ContainerMetricsCollector) Collect(ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ch); err != nil {
+		log.Error("failed collecting ContainerMetricsCollector metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+func (c *ContainerMetricsCollector) collect(ch chan<- prometheus.Metric) (*prometheus.Desc, error) {
+
+	// Types Container is passed to get the containers compute systems only
+	containers, err := hcsshim.GetContainers(hcsshim.ComputeSystemQuery{Types: []string{"Container"}})
+	if err != nil {
+		log.Error("Err in Getting containers:", err)
+		return nil, err
+	}
+
+	if len(containers) == 0 {
+		log.Warn("no containers on node")
+		return nil, nil
+	}
+
+	for _, containerDetails := range containers {
+		containerId := containerDetails.ID
+
+		container, err := hcsshim.OpenContainer(containerId)
+		if err != nil {
+			log.Error("err in opening container: ", containerId)
+			continue
+		}
+
+		cstats, err := container.Statistics()
+		if err != nil {
+			log.Error("err in fetching container Statistics: ", containerId)
+			continue
+		}
+		// HCS V1 is for docker runtime. Add the docker:// prefix on container_id
+		containerId = "docker://" + containerId
+
+		ch <- prometheus.MustNewConstMetric(
+			c.ContainerAvailable,
+			prometheus.CounterValue,
+			1,
+			containerId,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.UsageCommitBytes,
+			prometheus.GaugeValue,
+			float64(cstats.Memory.UsageCommitBytes),
+			containerId,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.UsageCommitPeakBytes,
+			prometheus.GaugeValue,
+			float64(cstats.Memory.UsageCommitPeakBytes),
+			containerId,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.UsagePrivateWorkingSetBytes,
+			prometheus.GaugeValue,
+			float64(cstats.Memory.UsagePrivateWorkingSetBytes),
+			containerId,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.TotalRuntime100ns,
+			prometheus.CounterValue,
+			float64(cstats.Processor.TotalRuntime100ns)*ticksToSecondsScaleFactor,
+			containerId,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.RuntimeUser100ns,
+			prometheus.CounterValue,
+			float64(cstats.Processor.RuntimeUser100ns)*ticksToSecondsScaleFactor,
+			containerId,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.RuntimeKernel100ns,
+			prometheus.CounterValue,
+			float64(cstats.Processor.RuntimeKernel100ns)*ticksToSecondsScaleFactor,
+			containerId,
+		)
+
+		if len(cstats.Network) == 0 {
+			log.Warn("No Network Stats for container: ", containerId)
```

Comment:
Does this indicate an invalid container configuration? Otherwise, like the warning for no containers above, I'd skip the warning and just possibly add a metric for the number of interfaces on the container.

Reply:
Yes. There should be network stats in normal scenario so having this log should be okay. I changed it to info instead of warning.